android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/BaseViewRouter.kt

```diff
@@ -0,0 +1,160 @@
+package com.badoo.common.rib
+
+import android.os.Bundle
+import android.os.Parcelable
+import android.util.SparseArray
+import android.view.ViewGroup
+import com.badoo.common.rib.requestcode.RequestCodeRegistry
+import com.badoo.common.rib.routing.action.RoutingAction
+import com.uber.rib.core.Interactor
+import com.uber.rib.core.RibView
+import com.uber.rib.core.Router
+import java.util.UUID
+
+@Suppress("FINITE_BOUNDS_VIOLATION_IN_JAVA")
+open class BaseViewRouter<V : RibView, I : Interactor<*>>(
+    private val viewFactory: ViewFactory<V>?,
+    interactor: I
+) : Router<I>(interactor) {
+    private val ribName: String = "${this::class.java.name}.${UUID.randomUUID()}"
+    private var ribId: Int? = null
+    internal var view: V? = null
+    protected var parentViewGroup: ViewGroup? = null
+
+    private lateinit var attachPermanentParts: RoutingAction<V>
+    protected open val permanentParts: List<() -> BaseViewRouter<*, *>> = emptyList()
+    private var savedViewState: SparseArray<Parcelable> = SparseArray()
+
+    private fun attachPermanentParts() {
+        permanentParts.forEach {
+            addChild(it()) // fixme save and restore these as well
+        }
+    }
+
+    override fun dispatchAttach(savedInstanceState: Bundle?, tag: String) {
+        super.dispatchAttach(savedInstanceState, tag)
+        updateRibId(savedInstanceState?.getInt(KEY_RIB_ID, generateGroupId()) ?: generateGroupId())
+        savedViewState = savedInstanceState?.getSparseParcelableArray<Parcelable>(KEY_VIEW_STATE) ?: SparseArray()
+        attachPermanentParts()
+    }
+
+    private fun generateGroupId(): Int =
+        requestCodeRegistry.generateGroupId(ribName)
+
+    private fun generateRequestCode(code: Int): Int =
+        requestCodeRegistry.generateRequestCode(ribName, code)
+
+    private fun updateRibId(value: Int) {
+        ribId = value
+    }
+
+    override fun saveInstanceState(outState: Bundle) {
+        super.saveInstanceState(outState)
+        outState.putInt(KEY_RIB_ID, ribId ?: generateGroupId().also { updateRibId(it) })
+        outState.putSparseParcelableArray(KEY_VIEW_STATE, savedViewState)
+    }
+
+    // todo move this to base router class so no casting is needed inside foreach
+    fun onAttachToView(parentViewGroup: ViewGroup) {
```

Comment:
I wouldn't call it `onAttachToView`, it actually **attaches** it.
So `attachView` probably?